scipy/linalg/_matfuncs_sqrtm.c

```diff
@@ -0,0 +1,1006 @@
+    // Pointers to the variables inside the buffer for easy access
+    // --------------------------------------------------------------------
+    float* restrict data = &buffer[0];
+    float* restrict data2 = &buffer[n*n];
+    float* restrict vs = &buffer[2*n*n];
+    float* restrict wr = &buffer[4*n*n];
+    float* restrict wi = &buffer[4*n*n + n];
+    float* restrict work = &buffer[4*n*n + 2*n];
+    // Cover the first 2n*n and vs area with a complex pointer just in case
+    SCIPY_C* complex_data = &((SCIPY_C*)buffer)[0];
+    SCIPY_C* complex_vs = &((SCIPY_C*)buffer)[n*n];
+
+
+    /*====================================================================
+    |                    MAIN nxn SLICE LOOP                             |
+    ====================================================================*/
+    for (npy_intp idx = 0; idx < outer_size; idx++) {
+        /*------------------------------------------------------------------
+        Copy the current slice into buffer in F-layout for LAPACK calls. Note
+        that the input can be in C or F layout and not necessarily contiguous.
+        Since the input is predominantly C contiguous, we copy the data in
+        C-layout first and then transpose it to F-layout.
+
+        The output is always a C contiguous array hence we visit slices in
+        C-order for easy insertion to the output array; by just offsetting
+        idx*n*n * sizeof(dtype). For this reason, we  start the offset
+        calculation from the inner-most outer dimension.
+        ------------------------------------------------------------------*/
+        npy_intp offset = 0;
+        npy_intp temp_idx = idx;
+        for (int i = ndim - 3; i >= 0; i--) {
+            offset += (temp_idx % shape[i]) * strides[i];
+            temp_idx /= shape[i];
+        }
+        float* restrict slice_ptr = (float*)(Am_data + (offset/sizeof(float)));
+
+        for (int i = 0; i < n; i++) {
+            for (int j = 0; j < n; j++) {
+                data2[i * n + j] = *(slice_ptr + (i*strides[ndim - 2]/sizeof(float)) + (j*strides[ndim - 1]/sizeof(float)));
+            }
+        }
+        swap_cf_s(data2, data, n, n, n);
+
+        isComplex = 0;
+        // ------------------------------------------------------------------------
+        // Check if array is (quasi)upper triangular.
+        // ------------------------------------------------------------------------
+        int isSchur = isschurf(data, n);
+
+        if (!isSchur)
+        {
+            sgees_("V", "N", NULL, &intn, data, &intn, &sdim, wr, wi, vs, &intn, work, &lwork, NULL, &info);
+            if (info != 0)
+            {
+                free(buffer);
+                *sq_info = -102;
+                return;
+            }
+        } else {
+            // Manually get the eigenvalues with possible 2x2 blocks on the diagonal
+            for (Py_ssize_t col = 0; col < n; col++)
```

Comment:
this should probably be `npy_intp` even if (assumingly) these types are the same (?). In other places indexing uses npy_intp.

Reply:
We don't need to use npy_intp for non NumPy entities such as malloc'd buffers etc. In fact we don't need to use npy_intp anywhere but I did to visually separate when I'm touching NumPy array stuff.